pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java

```diff
@@ -0,0 +1,114 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import org.apache.pinot.common.proto.Server;
+import org.apache.pinot.core.transport.ServerRoutingInstance;
+import org.apache.pinot.spi.config.table.TableType;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.assertTrue;
+
+
+public class StreamingReduceServiceTest {
+
+  @Test
+  public void testThreadExceptionTransfer() {
+    // simulate a thread exception in gRPC call and verify that the thread can transfer the exception
+    Iterator<Server.ServerResponse> mockedResponse = (Iterator<Server.ServerResponse>) mock(Iterator.class);
+    when(mockedResponse.hasNext()).thenReturn(true);
+    final String exceptionMessage = "Some exception";
+    final RuntimeException innerException = new RuntimeException(exceptionMessage);
+    when(mockedResponse.next()).thenThrow(innerException);
+    final ExecutorService threadPoolService = Executors.newFixedThreadPool(1);
+    final ServerRoutingInstance routingInstance =
+        new ServerRoutingInstance("localhost", 9527, TableType.OFFLINE, false);
+    // supposedly we can use TestNG's annotation like @Test(expectedExceptions = { IOException.class }) to verify
+    // here we hope to verify deeper to make sure the thrown exception is nested inside the exception
+    assertTrue(verifyException(() -> {
+          StreamingReduceService.processIterativeServerResponse(mock(StreamingReducer.class),
+              threadPoolService,
+              ImmutableMap.of(routingInstance, mockedResponse),
+              1000,
+              mock(BaseReduceService.ExecutionStatsAggregator.class));
+          return null;
+        }, (cause) -> cause.getMessage().contains(exceptionMessage))
+    );
+  }
+
+  @Test
+  public void testExecutionTimeout()
+      throws Exception {
+    // simulate a thread timeout in gRPC call and verify that the thread can transfer the exception
+    Iterator<Server.ServerResponse> mockedResponse = (Iterator<Server.ServerResponse>) mock(Iterator.class);
+    when(mockedResponse.hasNext()).thenReturn(true);
+    when(mockedResponse.next()).then(new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocationOnMock)
+          throws Throwable {
+        Thread.sleep(1000);
+        return null;
+      }
+    });
+    final ExecutorService threadPoolService = Executors.newFixedThreadPool(1);
+    final ServerRoutingInstance routingInstance =
+        new ServerRoutingInstance("localhost", 9527, TableType.OFFLINE, false);
+    // supposedly we can use TestNG's annotation like @Test(expectedExceptions = { IOException.class }) to verify
+    // here we hope to verify deeper to make sure the thrown exception is nested inside the exception
```

Comment:
It's hard to parse this sentence. I suggest changing it to something like "We can't use TestNG's expectedException because the exception we want to verify was thrown may be arbitrarily nested"

Reply:
Thanks. Rephrased